test/blackbox-tests/test-cases/watching/dir-target-promotion.t

```diff
@@ -0,0 +1,170 @@
+  $ build d1
+  Success
+  $ cat d1/a d1/b d1/d2/c
+  +++
+  $ ls d1 | grep unexpected
+  [1]
+  $ ls d1/d2 | grep unexpected
+  [1]
+
+We're done.
+
+  $ stop_dune
+  Success, waiting for filesystem changes...
+  Success, waiting for filesystem changes...
+  Success, waiting for filesystem changes...
+  Success, waiting for filesystem changes...
+  Success, waiting for filesystem changes...
+  File "dune", line 1, characters 0-151:
+  1 | (rule
+  2 |   (mode promote)
+  3 |   (deps src (sandbox always))
+  4 |   (targets (dir d1))
+  5 |   (action (bash "mkdir -p d1/d2; cp src d1/a; cp src d1/b; cp src d1/d2/c")))
+  Error: Cannot promote files to "d1/d2".
+  Reason: "d1/d2" is not a directory.
+  -> required by _build/default/d1
+  -> required by alias d1/all
+  -> required by alias d1/default
+  Had errors, waiting for filesystem changes...
+  Success, waiting for filesystem changes...
+  File "dune", line 1, characters 0-151:
+  1 | (rule
+  2 |   (mode promote)
+  3 |   (deps src (sandbox always))
+  4 |   (targets (dir d1))
+  5 |   (action (bash "mkdir -p d1/d2; cp src d1/a; cp src d1/b; cp src d1/d2/c")))
+  Error: This rule defines a directory target "d1" that matches the requested
+  path "d1/d2/unexpected-dir-2" but the rule's action didn't produce it
+  -> required by alias d1/d2/unexpected-dir-2/all
+  -> required by alias d1/d2/unexpected-dir-2/default
+  File "dune", line 1, characters 0-151:
+  1 | (rule
+  2 |   (mode promote)
+  3 |   (deps src (sandbox always))
+  4 |   (targets (dir d1))
+  5 |   (action (bash "mkdir -p d1/d2; cp src d1/a; cp src d1/b; cp src d1/d2/c")))
+  Error: This rule defines a directory target "d1" that matches the requested
+  path "d1/unexpected-dir-1" but the rule's action didn't produce it
+  -> required by alias d1/unexpected-dir-1/all
+  -> required by alias d1/unexpected-dir-1/default
+  Success, waiting for filesystem changes...
+
+Now test file-system events generated during directory target promotion.
+
+  $ rm -rf d1
+  $ start_dune --debug-cache=fs
+  $ build d1
+  Success
+
+  $ stop_dune > .#debug-output
+
+Show that Dune ignores the initial "dune-workspace" events (injected by Dune).
+
+  $ cat .#debug-output | grep dune-workspace
+  Updating dir_contents cache for "dune-workspace": Skipped
+  Updating path_digest cache for "dune-workspace": Skipped
+  Updating path_stat cache for "dune-workspace": Updated { changed = false }
+
+Dune correctly notices that the contents of . changed because [d1] was created.
+
+  $ cat .#debug-output | grep '"."'
+  Updating dir_contents cache for ".": Updated { changed = true }
+  Updating path_digest cache for ".": Skipped
+  Updating path_stat cache for ".": Updated { changed = false }
+
+Here [path_stat] of the directory [d1] changed, presumably, because its [mtime]
+changed. Ideally, Dune shouldn't trigger itself when creating files/directories
+inside promoted directories.
+
+# CR-someday amokhov: Fix this by not watching [mtime]s.
```

Comment:
Hmm, it looks like getting rid of `mtime`s didn't fix this test. I'd like to investigate a bit further.

Reply:
Ah, of course, we're still using `mtime` to compute the digest of a directory in `Cached_digest`. I am going to fix this in a separate PR.

Reply:
(... and, of course, it has also changed because previously `d1` didn't exist, so I'm going to remove the `CR-someday`.)